### ISSUE TITLE
Context support

### ISSUE THREAD

**Issue by GeertJohan** [lib/pq#786]
I'd like for pq to support the *Context calls.

In one of my projects I started using [github.com/basvanbeek/ocsql](https://github.com/basvanbeek/ocsql). It wraps  the driver and adds OpenCensus instrumentation (tracing and metrics). This works great, even when wrapping the sql.DB with [sqlx](https://github.com/jmoiron/sqlx) for extra functionality. However, it requires the driver to support Context calls as defined in the stdlib (BeginTx, ExecContext, etc.). Therefore, I (temporarily?) switched to [github.com/jackc/pgx/stdlib](https://godoc.org/github.com/jackc/pgx/stdlib

I think that pg should support these functions. https://golang.org/pkg/database/sql/driver

**Comment by GeertJohan** [comment]
Slightly related to #667

**Comment by Gunni** [comment]
I'd like to push for support aswell, i had a heck of a time debugging why the database connection has such an insanely long timeout when the database is unresponsive!

The database was also ignoring the timeout context it was being given, and with a default timeout of 42 seconds!!! (in my case a timeout of at MOST 1 second is plenty!!)

**Comment by basvanbeek** [comment]
`ocsql` will work regardless of `lib/pq` supporting *Context or not. `ocsql` needs you as consumer to call the *Context methods so it can extract the current SpanContext from context.Context. After that the driver can either support context.Context (as it could provide cancellation/deadline logic) or not at all. 

Long story short. Support for *Context would be highly appreciated for timeout but also expensive query cancellation in case upstream no longer cares for the query result.

**Comment by mjibson** [comment]
I'm not aware of a reason to use lib/pq over pgx. It's better supported, more correct, and is faster.

**Comment by GeertJohan** [comment]
@basvanbeek For me, ocsql was complaining that it was missing the Context* methods in the sql driver, and could not continue. Perhaps this was a bug or missing feature that has been implemented since I last tried it? I'm now using pgx so it's all fine now.

@mjibson `lib/pq` recovers when a connection is broken. If I restart my database, the first few calls using `pgx` fail with connection errors. Not nice. Also: I'm not sure if pgx is faster when used as stdlib driver for database/sql. Afaik the benchmark was done using pgx directly.

**Comment by cgilling** [comment]
I might be missing something, but isn't it implemented here? https://github.com/lib/pq/blob/0ac245bc3be505fde6900e1ca0f957c0845cbfa2/conn_go18.go#L13

**Comment by KillerX** [comment]
@cgilling I was wondering the same thing (we are using `*Context` pretty much exclusively), but for some reason `ocsql` looses the information about the parent span. The results appear in the traces (Stackdriver for us) but they are *stand alone* instead of connected to the relevant parent span.

This happens in the simplest cases, and even when a span is manually created just before the query, like in the example below.

```
func (db DatabaseHandler) GetStuff(ctx context.Context, someParam string) (*db.Row) {
	ctx, span := trace.StartSpan(ctx, "SomeQuery")
	defer span.End()

	row := db.PreparedQuery.QueryRowContext(ctx, someParam)

	return row
}
```

In this case the new span will be correctly correlated to the parent span but the query will be its own separate trace.

I have tested `pgx` as @mjibson suggested, and without changing anything else than the driver, the queries are correctly correlated.

This leads me to believe that there is some issue with how the context is handled, but I was at this point unable to discover what exactly is the issue.

**Comment by basvanbeek** [comment]
It seems lib/pq does not support driver.StmtQueryContext and driver.StmtExecContext. It only implements the Context methods on a driver.Conn.

ocsql tries to be transparent to sql/database on which enhancement interfaces are implemented. It hides interfaces not implemented by the underlying driver. In this case this means a userland call to stmt.QueryRowContext in database/sql is sent to driver's stmt.Query thus dropping the context.

Best would be for lib/pq to implement these functions.

**Comment by KillerX** [comment]
@basvanbeek thanks for the explanation. Turns out there is actually a pr pending for exactly this: #768

**Comment by montanaflynn** [comment]
Any update on this? The PR is still open.

**Comment by Gunni** [comment]
Does https://github.com/lib/pq/pull/839 fix this?

**Comment by KillerX** [comment]
@Gunni I don't think so since functions like `QueryContext` are still missing (at least as far as I can see).

**Comment by Gunni** [comment]
This bug is now 5 years old, starting primary school soon, I wonder if it will ever be fixed...

**Comment by gmbuell** [comment]
There is a remaining COPY-specific context gap on current master `10237a6`: `copyin` does not implement `driver.StmtExecContext`, so database/sql falls back to the legacy Exec method and a COPY buffer flush blocked in net.Conn.Write cannot observe the caller's deadline.

[`TestConnectionIONetworkRegressionCopyExecContextInterruptsBlockedFlush`](https://github.com/gmbuell/pq/blob/1c9dd44e4cd92b7b48838b20618e55617b40078f/connection_io_network_regression_test.go#L234) is a deterministic net.Pipe reproduction; it expects the exact context error and an unusable connection after interruption. The candidate implementation is in [this change](https://github.com/gmbuell/pq/commit/c334e3c0d5076b760ff6d66c332c681098c90662).
